tests/aklite_offline_test.cc

```diff
@@ -425,29 +429,75 @@ TEST_F(AkliteOffline, RollbackIfAppStartFailsWithAppShortlisting) {
+  ASSERT_TRUE(initial_target_.MatchTarget(getCurrent()));
+}
+
+TEST_F(AkliteOffline, RollbackToInitialTargetIfAppDrivenRolllback) {
+  const auto app01{createApp("app-01")};
+  preloadApps({app01}, {}, false);
+
+  // remove the current target app from the store/install source dir
+  boost::filesystem::remove_all(app_store_.appsDir());
+  const auto new_target{addTarget({createApp("app-01", "compose-start-failure")})};
+  ASSERT_EQ(install(), offline::PostInstallAction::NeedReboot);
+  reboot();
+  ASSERT_EQ(run(), offline::PostRunAction::RollbackNeedReboot);
+  reboot();
+  const auto current{getCurrent()};
+  ASSERT_TRUE(initial_target_.MatchTarget(getCurrent()));
+  ASSERT_EQ(Target::appsJson(current).size(), 1);
+  ASSERT_TRUE(Target::appsJson(current).isMember(app01.name));
+  ASSERT_EQ(Target::appsJson(current)[app01.name]["uri"], app01.uri);
+}
+
+TEST_F(AkliteOffline, RollbackToUnknown) {
+  const auto app01{createApp("app-01")};
+  preloadApps({app01}, {}, false);
+  // remove the current target app from the store/install source dir
+  boost::filesystem::remove_all(app_store_.appsDir());
+  const auto new_target{addTarget({createApp("app-01")})};
+
+  // make the initial Target setting fail, so the current Target is "unknown"
+  const Docker::Uri uri{Docker::Uri::parseUri(app01.uri)};
+  const auto app_dir{test_dir_.Path() / "reset-apps" / "apps" / uri.app / uri.digest.hash()};
+  Utils::writeFile(app_dir / Docker::Manifest::Filename, std::string("broken json"));
+
+  ASSERT_EQ(install(), offline::PostInstallAction::NeedReboot);
+  reboot();
+  // emulate "normal" rollback - boot on the previous target, which is "unknown"
+  sys_repo_.deploy(initial_target_.sha256Hash());
   ASSERT_EQ(run(), offline::PostRunAction::RollbackToUnknown);
-  ASSERT_EQ(getCurrent().sha256Hash(), initial_target_.sha256Hash());
+  const auto current{getCurrent()};
+  ASSERT_TRUE(Target::isUnknown(current));
+  ASSERT_EQ(current.sha256Hash(), initial_target_.sha256Hash());
 }
 
 TEST_F(AkliteOffline, RollbackToUnknownIfAppDrivenRolllback) {
-  preloadApps({createApp("app-01")}, {}, false);
+  const auto app01{createApp("app-01")};
+  preloadApps({app01}, {}, false);
 
   // remove the current target app from the store/install source dir
   boost::filesystem::remove_all(app_store_.appsDir());
   const auto new_target{addTarget({createApp("app-01", "compose-start-failure")})};
+
+  // make the initial Target setting fail, so the current Target is "unknown"
+  const Docker::Uri uri{Docker::Uri::parseUri(app01.uri)};
+  const auto app_dir{test_dir_.Path() / "reset-apps" / "apps" / uri.app / uri.digest.hash()};
+  Utils::writeFile(app_dir / Docker::Manifest::Filename, std::string("broken json"));
   ASSERT_EQ(install(), offline::PostInstallAction::NeedReboot);
   reboot();
   ASSERT_EQ(run(), offline::PostRunAction::RollbackToUnknownIfAppFailed);
+  // Cannot perform rollback to unknown, so still booted on a new target's hash
```

Comment:
I am sitting on a fence about this use-case. There are two options:

1. Leave it as it is. A device is booted on a new target's hash but cannot run the updated Apps. No rollback because the rollback Target is "unknown".

2. Re-deploy the rollback ostree deployment if found any (even if it's unknown), and return `RollbackNeedReboot`, so the following reboot makes a device to boot on the unknown Target's hash.

Reply:
It would be pretty hard to get into this situation, right? That kind of makes me lean towards going with 1 since its not requiring a reboot. Neither situation is great, but at least with option 1 you aren't rebooting a device

Reply:
I incline towards the option 1 too. The online aklite can recover from such state by applying the following update, and it's always better to stay on the newest ostree/rootfs to do so. When it comes to the offline update, I think it's up to a client to decide what to do in such the case, whether to rollback to the previous rootfs version or stay with the current one and try to start specific apps manually. I will document it.